Update and improve cache operations.

adamgreig opened the pull request:
Closes #47, #188.

I've implemented the proposed methods from #47 and marked all d-cache invalidation functions as unsafe. It's not unsafe to invalidate i-cache or branch predictor as they are read-only caches. The clean and clean+invalidate operations do not alter memory from the executing core's point of view so are also safe.

It wasn't possible to remove the requirement to pass in `&mut CPUID` as you require synchronized access to `CPUID` to read the number of sets and ways in the cache, which is required to fully clean or invalidate them, which is required to enable or disable them. So it goes.

Breaking change due to changing safety of d-cache invalidation functions.

rust-highfive commented:
r? @ithinuel

(rust_highfive has picked a reviewer for you, use r? to override)

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -518,21 +555,80 @@ impl SCB {
+    /// recent writes to `obj` to be lost, potentially including the write that initialised it.
+    /// Therefore, this method may cause uninitialised memory or invalid values to be read,
+    /// resulting in undefined behaviour. You must ensure that main memory contains a valid and
+    /// initialised value for T before invalidating `obj`.
```
Isn't it unsound in general when the memory a `&T` points to is modified (unless `T` is an `UnsafeCell`)?

adamgreig replied:
Do you think we should require `&mut` here instead of `&`, or just give a more dire warning about potential consequences?

jonas-schievink replied:
`&mut` sounds like a good option, as long as it doesn't make the API unusable. Do you happen to have any real-world examples where this API would be used?

adamgreig replied:
I've used the previous invalidate API in ethernet peripheral drivers where the descriptors and packet memory are shared between CPU and peripheral. In all of those instances the code had &mut access, though the receive buffers were in principle read-only.

The other use case I can see is sharing memory between cores on a multi-core device, e.g. the STM32H7 dual-core series have D-cache for the M7 core, so communicating with the M4 core could require cache management. Imagine a synchronisation primitive which needs to invalidate some interior memory to ensure it's reading a fresh value from the other core, and wants to be able to do this with only `&self`. But probably such a primitive should contain an `UnsafeCell` which can handle the interior mutability, and then it would still be able to pass &mut to the cache invalidation function.

On balance I think it's probably a more obvious API if it takes `&mut` to make it clear that the value might change, and in the rare case someone needs to access it with a non-mut, it's only a one line function that calls another public unsafe method anyway, so it's not that much extra effort. I'll update this PR.

adamgreig commented:
A couple of other notes:

* The comments refer to 32-byte alignment and length. Technically in ARMv7-M this length is core-specific and available from the DMinLine field of the CTR register. Currently, only Cortex-M7 cores can have a D-cache, and Cortex-M7 cores can only have 32-byte DMinLine. In the future ARM might release other cores with other cache sizes. I'll update the docstrings to clarify.
* The above means we probably should not have an assert that the memory being invalidated is 32-byte aligned and a multiple of 32 bytes long. Which is a shame as those asserts would probably be quire useful and make this a less sharp API. Perhaps we could read the DMinLine register and check based on that, I'll investigate.

Please don't merge just yet.

adamgreig commented:
I've updated the documentation to more clearly convey what's required and what will happen, updated the methods to take `&mut` where appropriate, and updated the underlying loops to use `DminLine` to obtain the cache size at runtime instead of hardcoding 32, although at the moment all released devices will use 32.

This is ready for review now.

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -486,88 +508,254 @@ impl SCB {
+                                          slice.len() * core::mem::size_of::<T>());
+    }
+
     /// Cleans D-cache by address
```
```suggestion
    /// Cleans D-cache by address.
```

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -486,88 +508,254 @@ impl SCB {
     ///
-    /// `addr`: the address to clean
-    /// `size`: size of the memory block, in number of bytes
+    /// * `addr`: the address to clean
```
```suggestion
    /// * `addr`: The address to start cleaning at.
```

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -486,88 +508,254 @@ impl SCB {
-    /// `addr`: the address to clean
-    /// `size`: size of the memory block, in number of bytes
+    /// * `addr`: the address to clean
+    /// * `size`: number of bytes to clean
```
```suggestion
    /// * `size`: The number of bytes to clean.
```

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -486,88 +508,254 @@ impl SCB {
         crate::asm::isb();
     }
 
+    /// Cleans an object in D-cache
```
```suggestion
    /// Cleans an object from the D-cache.
```

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -486,88 +508,254 @@ impl SCB {
 
+    /// Cleans an object in D-cache
+    ///
+    /// * `obj`: Object to clean
```
```suggestion
    /// * `obj`: The object to clean.
```

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -486,88 +508,254 @@ impl SCB {
+        self.clean_dcache_by_address(obj as *const T as usize, core::mem::size_of::<T>());
+    }
+
+    /// Cleans a slice in D-cache
```
```suggestion
    /// Cleans a slice from D-cache.
```

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -486,88 +508,254 @@ impl SCB {
+
+    /// Cleans a slice in D-cache
+    ///
+    /// * `slice`: Slice to clean
```
```suggestion
    /// * `slice`: The slice to clean.
```

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -486,88 +508,254 @@ impl SCB {
+                                     slice.len() * core::mem::size_of::<T>());
+    }
+
     /// Cleans and invalidates D-cache by address
```
```suggestion
    /// Cleans and invalidates D-cache by address.
```

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -486,88 +508,254 @@ impl SCB {
     ///
-    /// `addr`: the address to clean and invalidate
-    /// `size`: size of the memory block, in number of bytes
+    /// * `addr`: the address to clean and invalidate
```
```suggestion
    /// * `addr`: The address to clean and invalidate.
```

jonas-schievink commented on src/peripheral/scb.rs:
```diff
@@ -486,88 +508,254 @@ impl SCB {
-    /// `addr`: the address to clean and invalidate
-    /// `size`: size of the memory block, in number of bytes
+    /// * `addr`: the address to clean and invalidate
+    /// * `size`: number of bytes to clean and invalidate
```
```suggestion
    /// * `size`: The number of bytes to clean and invalidate.
```

adamgreig commented:
Thanks, I've applied suggestions and updated some other `initialised` :gb:

jonas-schievink commented:
Great, let's do this!

bors r+

jonas-schievink commented:
Apparently we're downloading stuff from Arm servers during CI. Not sure why but we shouldn't be doing that.

adamgreig commented:
bors retry

>

adamgreig commented:
We download the gcc toolchain which is used to verify the precompiled
binary blobs are correct compared to the source, iirc.

On Wed, 5 Feb 2020, 00:09 Jonas Schievink, <notifications@github.com> wrote:

> Apparently we're downloading stuff from Arm servers during CI. Not sure
> why but we shouldn't be doing that.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-embedded/cortex-m/pull/192?email_source=notifications&email_token=AAALQ4YKQ7345V63VXIDXILRBH7VHA5CNFSM4KHLQKG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKZVLBI#issuecomment-582178181>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALQ427ZFJOESIJ4OQBXWDRBH7VHANCNFSM4KHLQKGQ>
> .
>

jonas-schievink commented:
Using `apt-get` is better for that since Travis caches it. I'll open an issue.

adamgreig commented:
At the time I don't think a suitable version was available in apt and it
still might not be. We could have Travis cache the downloaded toolchain,
though.

On Wed, 5 Feb 2020, 00:16 Jonas Schievink, <notifications@github.com> wrote:

> Using apt-get is better for that since Travis caches it. I'll open an
> issue.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-embedded/cortex-m/pull/192?email_source=notifications&email_token=AAALQ46FXWJKHYLJTOSA3QTRBIANJA5CNFSM4KHLQKG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKZVZOY#issuecomment-582180027>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALQ4ZGRH6657UABFACWDDRBIANJANCNFSM4KHLQKGQ>
> .
>
